File: haskell/y2023/d04/Y2023D04.hs
module Main where

import Data.List (elem, null)
import Data.List.Split (splitOn)
import Data.String (words)
import Input (readLines)

main :: IO ()
main = do
  lines <- readLines 2023 4
  let cards = map parseCard lines
  p1 <- part1 cards
  putStrLn ("Part 1: " ++ show p1)
  p2 <- part2 cards
  putStrLn ("Part 2: " ++ show p2)

data Card = MkCard Int [Int] [Int]
  deriving (Show)

parseCard :: String -> Card
parseCard s = MkCard id winning numbers
  where
    split = splitOn ":" s
    id = read (drop 4 (head split))
    numsSplit = splitOn "|" (last split)
    winning = map read (words (head numsSplit))
    numbers = map read (words (last numsSplit))

part1 :: [Card] -> IO Int
part1 cards = do
  let scores = map scoreCard cards
  return (sum scores)

scoreCard :: Card -> Int
scoreCard card
  | winCount == 0 = 0
  | otherwise = 2 ^ (winCount - 1)
  where
    winCount = winCountCard card

winCountCard :: Card -> Int
winCountCard (MkCard _ w n) = length (filter (`elem` w) n)

part2 :: [Card] -> IO Int
part2 cards = do
  let counts = map (const 1) cards
  return $ part2Helper cards counts

part2Helper :: [Card] -> [Int] -> Int
part2Helper [] [] = 0
part2Helper (card : cards) (count : counts) = count + part2Helper cards newCounts
  where
    winCount = winCountCard card
    newCounts = map (+ count) (take winCount counts) ++ drop winCount counts

File: prolog/y2023/d04/y2023d04.pl
:- include('../../lib/input.pl').
:- include('../../lib/lineparser.pl').

main :-
    read_lines(Lines),
    maplist(parseCard, Lines, Cards),
    part1(Cards, P1),
    write("Part 1: "), write(P1), nl,
    part2(Cards, P2),
    write("Part 2: "), write(P2), nl.

parseCard(Line, card(Id, Winning, Numbers)) :-
    parseLine(
        [
            literal("Card"),
            whitespace, 
            capture(number), 
            literal(":"),
            whitespace, 
            capture(
                splitOn(
                    number, 
                    whitespace
                )
            ),
            whitespace,
            literal("|"),
            whitespace,
            capture(
                splitOn(
                    number, 
                    whitespace
                )
            )
        ], 
        Line,
        [], 
        [Id, Winning, Numbers]
    ).

% Part 1

part1([], 0).
part1([card(_, Winning, Numbers)|Cards], Res) :-
    intersection(Winning, Numbers, WinningNumbers),
    length(WinningNumbers, N),
    part1(Cards, Sum),
    ( N = 0 ->
        Res is Sum
    ;
        Res is Sum + (2 ^ (N - 1))
    ).

% Part 2

part2(Cards, Res) :-
    length(Cards, Len),
    length(Counts, Len),
    maplist(=(1), Counts),
    part2Helper(Cards, Counts, Res).

part2Helper([], [], 0).
part2Helper([card(_, Winning, Numbers)|Cards], [Count|Counts], Res) :-
    intersection(Winning, Numbers, WinningNumbers),
    length(WinningNumbers, N),
    increaseCounts(Count, N, Counts, NewCounts),
    part2Helper(Cards, NewCounts, SubRes),
    Res is Count + SubRes.

increaseCounts(_, 0, Counts, Counts).
increaseCounts(_, N, [], []) :- 
    N > 0.
increaseCounts(X, N, [Count|Counts], [NewCount|NewCounts]) :-
    N > 0,
    NewN is N - 1,
    NewCount is Count + X,
    increaseCounts(X, NewN, Counts, NewCounts).
